### ISSUE TITLE
XAML parser does not handle whitespace correctly

### ISSUE THREAD

**Issue by benstevens48** [microsoft/microsoft-ui-xaml#2886]
**Describe the bug**
I'm trying to set the content of a string property in XAML. Because it's a long string, I want to set is as node content not as an attribute. I would like to be able to place line breaks (that are preserved during parsing) in the string because it will eventually be rendered in a text block. So I tried putting the entity `&#xA;` into the content. However, it seems this is simply either ignored or interpreted as a space. This behaviour is incorrect as far as I can see. It should be interpreted as a line feed. I tried using an attribute instead, and then `&#xA;` is correctly interpreted as a line feed.

Also, I tried using the `xml:space="preserve"` attribute, but this simply causes an error on build. This should work according to the docs, although I am less concerned about this as this would force me to use very long lines so I don't really want to use this method.

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| November 2019 Update (18363)       | Yes   |

Not using WinUI.

**Comment by chingucoding** [comment]
Can you share a code sample? The XAML Controls Gallery uses xml:space="preserve" for the samples and those work correctly in the app.

**Comment by benstevens48** [comment]
The first part of the bug seems to apply to all situations. The second part seems to apply to setting the property of a custom class e.g.
```
<ResourceDictionary
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:App1">
    <local:Class1 x:Key="test" xml:space="preserve">
        <local:Class1.TestString>
            test
            test      
        </local:Class1.TestString>
    </local:Class1>
</ResourceDictionary>

```
where Class1 is just a plain C# class with a plain string property TestString.

**Comment by chingucoding** [comment]
I think you need to set xml:space="preserve" on the string property.

**Comment by benstevens48** [comment]
That works, if you remove the xml:space="preserve" from local:Class1. However, it's still a bug I think because it should be inherited and not cause an error.

**Comment by chingucoding** [comment]
Seems a bit confusing, though not to surprising in my opinion. The first point is definitely a bug.

**Comment by ranjeshj** [comment]
@alwu-msft @stevenbrix FYI

**Comment by MikeHillberg** [comment]
Agreed, the xml:space should be fine either at the Class tag or the String tag (this special attribute has scope rules).

That's an XML rule. XAML has special rules for content (as opposed to attributes) in addition to the XML rules. I don't remember the details (they're described in the [MS-XAML](https://docs.microsoft.com/en-us/openspecs/microsoft_domain_specific_languages/ms-xaml/fd13e8ed-dd75-4767-92cf-e418d2c39817?redirectedfrom=MSDN#:~:text=%5BMS-XAML%5D%3A%20XAML%20Object%20Mapping%20Specification%202006%201%20Technical,...%205%20Trademarks.%20...%206%20Fictitious%20Names.%20) spec), but basically runs of whitespace collapse to a single space. They generally follow the same content rules as HTML, and in this case Markdown is similar. 

We're working on adding CDATA support in [WinUI3](http://aka.ms/winui3), which will be a way to override this behavior.

**Comment by benstevens48** [comment]
OK, I understand that line feed characters placed directly in the content will be collapsed. I don't really understand why an XML entity like `&#xA;` needs to be collapsed. Is this really deliberate or is it just a consequence of the order in which things are processed? I can't think of a case where you would want to collapse a raw markup string of `&#xA;` into a space. The relevant part of the specification is here - https://docs.microsoft.com/en-us/openspecs/microsoft_domain_specific_languages/ms-xaml/cfe7ca97-3085-4363-af08-e01df1d586a4 - but I can't tell whether this applies to characters placed directly in the content or to XML entities as well. Perhaps there's some subtle reason for this collapsing behavior like needing to be able to output back to XML after the entities have be decoded, although I don't see when this wouldn't apply to attributes as well...

**Comment by MikeHillberg** [comment]
Xml entities are resolved by the Xml parser before Xaml processing begins. So Xaml can't actually tell if a line feed in content comes from a `\n` or a `&#xA;`.

**Comment by benstevens48** [comment]
Ah, I see. I suppose it would be possible to build the whitespace normalization into the XML reader step, but then you'd have a non-standard XML reader, which would probably be a bad idea. I also see now why `xml:space` might cause an issue - I thought this was just an XML-reader level directive, but now I see that it applies to both the XML-reader and to the whitespace normalization performed on the output of the XML-reader, so the XAML processor also needs to understand the `xml:space` directive, and I guess there are some holes in it.

**Comment by benstevens48** [comment]
But now I don't understand why `&#xA;` entities are not collapsed in attributes as well. I guess XAML is using an old non-strictly conformant XML processor. According to <https://docs.microsoft.com/en-us/openspecs/ie_standards/ms-xml/389b8ef1-e19e-40ac-80de-eec2cd0c58ae>, the XAML XML parser seems not to be using the 'NewParser' option, and doesn't conform to the official XML spec.

What is the best future-proof thing to do here? Is `xml:space` OK to use? It's been deprecated in HTML...

**Comment by alwu-msft** [comment]
Incidentally, XAML's attribute value normalization behavior is something I looked into at the beginning of this year. Basically, XAML's behavior here *is* in line with the standard. Because no DTD is presented for validation, attribute values (with the exception of well-known attributes like `xml:space`) are considered to be `string` typed (i.e. CDATA) and normalized accordingly (see [section 3.3.3 of the XML spec](https://www.w3.org/TR/xml/#AVNormalize) for details on how that works).

**Comment by benstevens48** [comment]
Oh, yes, I didn't appreciate the difference between character and entity references. If you had an entity reference to a line feed character that would be replaced be a space, but not a character reference!

**Comment by benstevens48** [comment]
Did the xml:space="preserve" get addressed in WinUI3? (The rest of the issue was mostly me not understanding the XML spec/the spec being confusing).
